File: java/src/main/java/com/fish/tree/Q1457.java
package com.fish.tree;

import java.util.LinkedList;
import java.util.List;

/**
 * @author liuqi
 * @date 2023/11/25
 */
public class Q1457 {
    class Solution {
        int ans = 0;

        public void help(TreeNode root, int[] cnt) {
            if (root == null) {
                return;
            }
            cnt[root.val]++;
            if (root.left == null && root.right == null) {
                int oddCnt = 0;
                for (int i = 0; i < 10; i++) {
                    if (cnt[i] % 2 == 1) {
                        oddCnt++;
                    }
                }
                ans += (oddCnt <= 1 ? 1 : 0);
                cnt[root.val]--;
                return;
            }
            help(root.left, cnt);
            help(root.right, cnt);
            cnt[root.val]--;
        }

        public int pseudoPalindromicPaths(TreeNode root) {
            help(root, new int[10]);
            return ans;
        }
    }
}

File: python/Q1457.py
# Definition for a binary tree node.
from typing import Optional
from TreeNode import TreeNode


class Solution:
    def __init__(self):
        self.ans = 0
        self.cnt = [0 for i in range(0, 10)]

    def help(self, root: TreeNode):
        if not root:
            return
        self.cnt[root.val] += 1
        if root.left is None and root.right is None:
            oddCnt = 0
            for i in self.cnt:
                if i % 2 == 1:
                    oddCnt += 1
            if oddCnt <= 1:
                self.ans += 1
                self.cnt[root.val] -= 1
                return
        self.help(root.left)
        self.help(root.right)
        self.cnt[root.val] -= 1

    def pseudoPalindromicPaths(self, root: Optional[TreeNode]) -> int:
        self.help(root)
        return self.ans


Solution().pseudoPalindromicPaths(TreeNode())
